File: hdf-generators/go/delta_test.go
package generators

import (
	"testing"

	hdf "github.com/mitre/hdf-libs/hdf-schema/dist/go/v3"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func makeTestBaseline(name string, reqs []hdf.BaselineRequirement) hdf.HDFBaseline {
	return hdf.HDFBaseline{
		Name:         name,
		Requirements: reqs,
	}
}

func makeTestReq(id string, title string) hdf.BaselineRequirement {
	return hdf.BaselineRequirement{
		ID:     id,
		Impact: 0.5,
		Title:  &title,
		Tags:   map[string]any{},
		Descriptions: []hdf.Description{
			{Label: "default", Data: "Test requirement"},
		},
	}
}

func TestGenerateDelta_PreservesOldCode(t *testing.T) {
	baseline := makeTestBaseline("test-profile", []hdf.BaselineRequirement{
		makeTestReq("SV-001", "Updated title"),
	})
	links := []LinkRecord{
		{OldID: "V-001", NewID: "SV-001", MatchMethod: "srgDeterministic",
			Confidence: 1.0, Relationship: "primary"},
	}
	codeMap := map[string]string{
		"V-001": "  describe command('audit') do\n    its('stdout') { should match /enabled/ }\n  end",
	}

	result := GenerateDelta(baseline, links, codeMap, nil, 1)

	require.Len(t, result.Profile.Controls, 1)
	control := result.Profile.Controls["controls/SV-001.rb"]
	assert.Contains(t, control, "control 'SV-001' do")
	assert.Contains(t, control, "Updated title")
	assert.Contains(t, control, "describe command('audit')")
	assert.NotContains(t, control, "TODO")
}

func TestGenerateDelta_GeneratesStubForUnmatched(t *testing.T) {
	baseline := makeTestBaseline("test-profile", []hdf.BaselineRequirement{
		makeTestReq("SV-002", "New control"),
	})
	links := []LinkRecord{
		{NewID: "SV-002", MatchMethod: "none", Relationship: "no-match"},
	}

	result := GenerateDelta(baseline, links, map[string]string{}, nil, 0)

	control := result.Profile.Controls["controls/SV-002.rb"]
	assert.Contains(t, control, "TODO")
}

func TestGenerateDelta_RelatedUsesOldCode(t *testing.T) {
	baseline := makeTestBaseline("test-profile", []hdf.BaselineRequirement{
		makeTestReq("SV-001", "Primary control"),
		makeTestReq("SV-002", "Related control"),
	})
	links := []LinkRecord{
		{OldID: "V-001", NewID: "SV-001", MatchMethod: "srgDeterministic",
			Confidence: 1.0, Relationship: "primary"},
		{OldID: "V-001", NewID: "SV-002", MatchMethod: "srgDeterministic",
			Confidence: 1.0, Relationship: "related"},
	}
	codeMap := map[string]string{
		"V-001": "  describe service('sshd') do\n    it { should be_running }\n  end",
	}

	result := GenerateDelta(baseline, links, codeMap, nil, 1)

	require.Len(t, result.Profile.Controls, 2)
	assert.Contains(t, result.Profile.Controls["controls/SV-001.rb"], "describe service('sshd')")
	assert.Contains(t, result.Profile.Controls["controls/SV-002.rb"], "describe service('sshd')")
}

func TestGenerateDelta_Statistics(t *testing.T) {
	baseline := makeTestBaseline("test-profile", []hdf.BaselineRequirement{
		makeTestReq("SV-001", "C1"),
		makeTestReq("SV-002", "C2"),
		makeTestReq("SV-003", "C3"),
	})
	links := []LinkRecord{
		{OldID: "V-001", NewID: "SV-001", MatchMethod: "srgDeterministic",
			Confidence: 1.0, Relationship: "primary"},
		{OldID: "V-001", NewID: "SV-002", MatchMethod: "srgDeterministic",
			Confidence: 1.0, Relationship: "related"},
		{NewID: "SV-003", MatchMethod: "none", Relationship: "no-match"},
	}

	result := GenerateDelta(baseline, links, map[string]string{"V-001": "# code"}, nil, 2)

	assert.Equal(t, 3, result.Statistics.NewControlsLength)
	assert.Equal(t, 2, result.Statistics.OldControlsLength)
	assert.Equal(t, 1, result.Statistics.Match)
	assert.Equal(t, 1, result.Statistics.DupMatch)
	assert.Equal(t, 1, result.Statistics.NoMatch)
	assert.Equal(t, 0, result.Statistics.PosMisMatch)
	assert.Equal(t, 2, result.Statistics.TotalMappedControls)
	// Verify invariant: totalMapped + noMatch = newControlsLength
	assert.Equal(t, result.Statistics.NewControlsLength,
		result.Statistics.TotalMappedControls+result.Statistics.NoMatch)
}

func TestGenerateDelta_EmptyMatchResult(t *testing.T) {
	baseline := makeTestBaseline("test-profile", []hdf.BaselineRequirement{
		makeTestReq("SV-001", "C1"),
		makeTestReq("SV-002", "C2"),
	})

	result := GenerateDelta(baseline, nil, map[string]string{}, nil, 0)

	assert.Len(t, result.Profile.Controls, 2)
	assert.Equal(t, 2, result.Statistics.NewControlsLength)
	assert.Equal(t, 0, result.Statistics.Match)
}

func TestGenerateDelta_NoCodeOption(t *testing.T) {
	baseline := makeTestBaseline("test-profile", []hdf.BaselineRequirement{
		makeTestReq("SV-001", "C1"),
	})
	links := []LinkRecord{
		{OldID: "V-001", NewID: "SV-001", MatchMethod: "exactId",
			Confidence: 1.0, Relationship: "primary"},
	}
	codeMap := map[string]string{"V-001": "  describe command('test') do\n  end"}

	result := GenerateDelta(baseline, links, codeMap, &DeltaOptions{NoCode: true}, 1)

	control := result.Profile.Controls["controls/SV-001.rb"]
	assert.Contains(t, control, "TODO")
	assert.NotContains(t, control, "describe command")
}

func TestGenerateDelta_SingleFile(t *testing.T) {
	baseline := makeTestBaseline("test-profile", []hdf.BaselineRequirement{
		makeTestReq("SV-001", "C1"),
		makeTestReq("SV-002", "C2"),
	})

	result := GenerateDelta(baseline, nil, map[string]string{}, &DeltaOptions{SingleFile: true}, 0)

	require.Len(t, result.Profile.Controls, 1)
	content := result.Profile.Controls["controls/controls.rb"]
	assert.Contains(t, content, "control 'SV-001'")
	assert.Contains(t, content, "control 'SV-002'")
}

func TestGenerateDelta_InspecYml(t *testing.T) {
	title := "Updated STIG Profile"
	baseline := hdf.HDFBaseline{
		Name:         "updated-stig",
		Title:        &title,
		Requirements: []hdf.BaselineRequirement{makeTestReq("SV-001", "C1")},
	}

	result := GenerateDelta(baseline, nil, map[string]string{}, nil, 0)

	assert.Contains(t, result.Profile.InSpecYml, "name: updated-stig")
	assert.Contains(t, result.Profile.InSpecYml, "title: Updated STIG Profile")
}

func TestGenerateUpgrade_DropsUnmatchedCurrentByDefault(t *testing.T) {
	// Default behavior (KeepUnmatched=false): if a current requirement has
	// no upstream match (DISA dropped it), it's dropped from the output.
	// Matches SAF CLI delta semantics.
	current := makeTestBaseline("current", []hdf.BaselineRequirement{
		makeTestReq("SV-001", "Matched control"),
		makeTestReq("SV-099", "Deprecated control — no upstream match"),
	})
	upstream := makeTestBaseline("upstream", []hdf.BaselineRequirement{
		makeTestReq("SV-001", "Matched control (upstream)"),
	})
	links := []LinkRecord{
		{OldID: "SV-001", NewID: "SV-001", MatchMethod: "srgDeterministic",
			Confidence: 1.0, Relationship: "primary"},
	}

	result := GenerateUpgrade(current, upstream, links, &UpgradeOptions{})

	ids := make([]string, 0, len(result.Baseline.Requirements))
	for _, r := range result.Baseline.Requirements {
		ids = append(ids, r.ID)
	}
	assert.Contains(t, ids, "SV-001", "matched control should be present")
	assert.NotContains(t, ids, "SV-099", "unmatched current should be dropped by default")
}

func TestGenerateUpgrade_KeepUnmatched_PreservesUnmatchedCurrent(t *testing.T) {
	// With KeepUnmatched=true, unmatched current requirements survive the
	// upgrade. Useful for users carrying custom controls outside the DISA
	// STIG, or who want to inspect what got dropped before committing.
	current := makeTestBaseline("current", []hdf.BaselineRequirement{
		makeTestReq("SV-001", "Matched control"),
		makeTestReq("SV-099", "Deprecated control — no upstream match"),
	})
	upstream := makeTestBaseline("upstream", []hdf.BaselineRequirement{
		makeTestReq("SV-001", "Matched control (upstream)"),
	})
	links := []LinkRecord{
		{OldID: "SV-001", NewID: "SV-001", MatchMethod: "srgDeterministic",
			Confidence: 1.0, Relationship: "primary"},
	}

	result := GenerateUpgrade(current, upstream, links, &UpgradeOptions{KeepUnmatched: true})

	ids := make([]string, 0, len(result.Baseline.Requirements))
	for _, r := range result.Baseline.Requirements {
		ids = append(ids, r.ID)
	}
	assert.Contains(t, ids, "SV-001", "matched control should be present")
	assert.Contains(t, ids, "SV-099", "unmatched current should be preserved with --keep-unmatched")
}

File: hdf-generators/test/delta.test.ts
import { describe, it, expect } from 'vitest';
import { generateDelta, generateUpgrade } from '../src/delta.js';
import type { LinkRecord, DeltaOptions } from '../src/delta-types.js';
import { makeBaseline, makeRequirement } from './helpers.js';

describe('generateDelta', () => {
  it('should preserve old code for matched requirements', () => {
    const newBaseline = makeBaseline({
      name: 'test-profile',
      requirements: [
        makeRequirement({
          id: 'SV-001',
          title: 'Updated title',
          impact: 0.7,
        }),
      ],
    });

    const linkRecords: LinkRecord[] = [
      {
        oldId: 'V-001',
        newId: 'SV-001',
        matchMethod: 'srgDeterministic',
        confidence: 1.0,
        relationship: 'primary',
        potentialMismatch: false,
      },
    ];

    const oldCodeMap = new Map([['V-001', "  describe command('audit') do\n    its('stdout') { should match /enabled/ }\n  end"]]);

    const result = generateDelta(newBaseline, linkRecords, oldCodeMap, undefined, 1);

    expect(result.profile.controls.size).toBe(1);
    const control = result.profile.controls.get('controls/SV-001.rb')!;
    expect(control).toContain("control 'SV-001' do");
    expect(control).toContain('Updated title');
    expect(control).toContain("describe command('audit')");
    expect(control).not.toContain('TODO');
  });

  it('should generate stubs for unmatched requirements', () => {
    const newBaseline = makeBaseline({
      name: 'test-profile',
      requirements: [
        makeRequirement({ id: 'SV-002', title: 'New control' }),
      ],
    });

    const linkRecords: LinkRecord[] = [
      {
        oldId: null,
        newId: 'SV-002',
        matchMethod: 'none',
        confidence: 0,
        relationship: 'no-match',
        potentialMismatch: false,
      },
    ];

    const result = generateDelta(newBaseline, linkRecords, new Map(), undefined, 0);

    const control = result.profile.controls.get('controls/SV-002.rb')!;
    expect(control).toContain('TODO');
    expect(control).not.toContain("describe command");
  });

  it('should use old code for related matches', () => {
    const newBaseline = makeBaseline({
      name: 'test-profile',
      requirements: [
        makeRequirement({ id: 'SV-001', title: 'Primary control' }),
        makeRequirement({ id: 'SV-002', title: 'Related control' }),
      ],
    });

    const linkRecords: LinkRecord[] = [
      {
        oldId: 'V-001',
        newId: 'SV-001',
        matchMethod: 'srgDeterministic',
        confidence: 1.0,
        relationship: 'primary',
        potentialMismatch: false,
      },
      {
        oldId: 'V-001',
        newId: 'SV-002',
        matchMethod: 'srgDeterministic',
        confidence: 1.0,
        relationship: 'related',
        potentialMismatch: false,
      },
    ];

    const oldCodeMap = new Map([['V-001', '  describe service("sshd") do\n    it { should be_running }\n  end']]);

    const result = generateDelta(newBaseline, linkRecords, oldCodeMap, undefined, 1);

    expect(result.profile.controls.size).toBe(2);
    const c1 = result.profile.controls.get('controls/SV-001.rb')!;
    const c2 = result.profile.controls.get('controls/SV-002.rb')!;
    expect(c1).toContain('describe service("sshd")');
    expect(c2).toContain('describe service("sshd")');
  });

  it('should compute SAF CLI-compatible statistics', () => {
    const newBaseline = makeBaseline({
      name: 'test-profile',
      requirements: [
        makeRequirement({ id: 'SV-001' }),
        makeRequirement({ id: 'SV-002' }),
        makeRequirement({ id: 'SV-003' }),
      ],
    });

    const linkRecords: LinkRecord[] = [
      {
        oldId: 'V-001', newId: 'SV-001', matchMethod: 'srgDeterministic',
        confidence: 1.0, relationship: 'primary', potentialMismatch: false,
      },
      {
        oldId: 'V-001', newId: 'SV-002', matchMethod: 'srgDeterministic',
        confidence: 1.0, relationship: 'related', potentialMismatch: false,
      },
      {
        oldId: null, newId: 'SV-003', matchMethod: 'none',
        confidence: 0, relationship: 'no-match', potentialMismatch: false,
      },
    ];

    const result = generateDelta(newBaseline, linkRecords, new Map([['V-001', '  # code']]), undefined, 2);

    expect(result.statistics.newControlsLength).toBe(3);
    expect(result.statistics.oldControlsLength).toBe(2);
    expect(result.statistics.match).toBe(1);
    expect(result.statistics.dupMatch).toBe(1);
    expect(result.statistics.noMatch).toBe(1);
    expect(result.statistics.posMisMatch).toBe(0);
    expect(result.statistics.totalMappedControls).toBe(2);
    // Invariant: totalMapped + noMatch = newControlsLength
    expect(result.statistics.totalMappedControls + result.statistics.noMatch)
      .toBe(result.statistics.newControlsLength);
  });

  it('should handle empty match result', () => {
    const newBaseline = makeBaseline({
      name: 'test-profile',
      requirements: [
        makeRequirement({ id: 'SV-001' }),
        makeRequirement({ id: 'SV-002' }),
      ],
    });

    const result = generateDelta(newBaseline, [], new Map(), undefined, 0);

    expect(result.profile.controls.size).toBe(2);
    expect(result.statistics.newControlsLength).toBe(2);
    expect(result.statistics.match).toBe(0);
  });

  it('should respect noCode option', () => {
    const newBaseline = makeBaseline({
      name: 'test-profile',
      requirements: [makeRequirement({ id: 'SV-001' })],
    });

    const linkRecords: LinkRecord[] = [
      {
        oldId: 'V-001', newId: 'SV-001', matchMethod: 'exactId',
        confidence: 1.0, relationship: 'primary', potentialMismatch: false,
      },
    ];
    const oldCodeMap = new Map([['V-001', '  describe command("test") do\n  end']]);

    const opts: DeltaOptions = { noCode: true };
    const result = generateDelta(newBaseline, linkRecords, oldCodeMap, opts, 1);

    const control = result.profile.controls.get('controls/SV-001.rb')!;
    expect(control).toContain('TODO');
    expect(control).not.toContain('describe command');
  });

  it('should generate valid inspec.yml from new baseline', () => {
    const newBaseline = makeBaseline({
      name: 'updated-stig',
      requirements: [makeRequirement({ id: 'SV-001' })],
      title: 'Updated STIG Profile',
    });

    const result = generateDelta(newBaseline, [], new Map(), undefined, 0);

    expect(result.profile.inspecYml).toContain('name: updated-stig');
    expect(result.profile.inspecYml).toContain('title: Updated STIG Profile');
  });

  it('drops unmatched-current requirements by default (matches SAF semantics)', () => {
    const current = makeBaseline({
      name: 'current',
      requirements: [
        makeRequirement({ id: 'SV-001' }),
        makeRequirement({ id: 'SV-099' }),
      ],
    });
    const upstream = makeBaseline({
      name: 'upstream',
      requirements: [makeRequirement({ id: 'SV-001' })],
    });
    const links: LinkRecord[] = [
      { oldId: 'SV-001', newId: 'SV-001', matchMethod: 'srgDeterministic',
        confidence: 1.0, relationship: 'primary', potentialMismatch: false },
    ];

    const result = generateUpgrade(current, upstream, links, {});

    const ids = result.baseline.requirements.map((r) => r.id);
    expect(ids).toContain('SV-001');
    expect(ids).not.toContain('SV-099');
  });

  it('preserves unmatched-current when keepUnmatched is true', () => {
    const current = makeBaseline({
      name: 'current',
      requirements: [
        makeRequirement({ id: 'SV-001' }),
        makeRequirement({ id: 'SV-099' }),
      ],
    });
    const upstream = makeBaseline({
      name: 'upstream',
      requirements: [makeRequirement({ id: 'SV-001' })],
    });
    const links: LinkRecord[] = [
      { oldId: 'SV-001', newId: 'SV-001', matchMethod: 'srgDeterministic',
        confidence: 1.0, relationship: 'primary', potentialMismatch: false },
    ];

    const result = generateUpgrade(current, upstream, links, { keepUnmatched: true });

    const ids = result.baseline.requirements.map((r) => r.id);
    expect(ids).toContain('SV-001');
    expect(ids).toContain('SV-099');
  });

  it('counts links with potentialMismatch into posMisMatch (not match)', () => {
    const current = makeBaseline({
      name: 'current',
      requirements: [makeRequirement({ id: 'SV-001' })],
    });
    const upstream = makeBaseline({
      name: 'upstream',
      requirements: [makeRequirement({ id: 'SV-001' })],
    });
    const links: LinkRecord[] = [
      { oldId: 'SV-001', newId: 'SV-001', matchMethod: 'srgCciTiebreak',
        confidence: 0.55, relationship: 'primary', potentialMismatch: true },
    ];

    const result = generateUpgrade(current, upstream, links, {});

    expect(result.statistics.posMisMatch).toBe(1);
    expect(result.statistics.match).toBe(0);
  });

  it('should handle singleFile option', () => {
    const newBaseline = makeBaseline({
      name: 'test-profile',
      requirements: [
        makeRequirement({ id: 'SV-001' }),
        makeRequirement({ id: 'SV-002' }),
      ],
    });

    const result = generateDelta(newBaseline, [], new Map(), { singleFile: true }, 0);

    expect(result.profile.controls.size).toBe(1);
    expect(result.profile.controls.has('controls/controls.rb')).toBe(true);
    const content = result.profile.controls.get('controls/controls.rb')!;
    expect(content).toContain("control 'SV-001'");
    expect(content).toContain("control 'SV-002'");
  });
});
